signature/jws/spec.go

```diff
@@ -4,46 +4,97 @@
 package jws
 
 import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"errors"
+	"fmt"
 	"time"
 
 	"github.com/golang-jwt/jwt/v4"
 	"github.com/notaryproject/notation-go"
+	"github.com/notaryproject/notation-go/spec/v1/signature"
 )
 
-// unprotectedHeader contains the header parameters that are not integrity protected.
-type unprotectedHeader struct {
-	TimeStampToken []byte   `json:"timestamp,omitempty"`
-	CertChain      [][]byte `json:"x5c,omitempty"`
-}
-
-// MediaTypeNotationPayload describes the media type of the payload of notation signature.
-const MediaTypeNotationPayload = "application/vnd.cncf.notary.v2.jws.v1"
-
-// payload contains the subject manifest and other attributes that have to be integrity
-// protected.
-type payload struct {
-	Notation notationClaim `json:"notary"`
+type notaryClaim struct {
 	jwt.RegisteredClaims
-}
-
-// notationClaim is the top level node and private claim, encapsulating the notary v2 data.
-type notationClaim struct {
-	Subject notation.Descriptor `json:"subject"`
+	Notary signature.JWSNotaryClaim `json:"notary"`
 }
 
 // packPayload generates JWS payload according the signing content and options.
-func packPayload(desc notation.Descriptor, opts notation.SignOptions) *payload {
+func packPayload(desc signature.Descriptor, opts notation.SignOptions) jwt.Claims {
 	var expiresAt *jwt.NumericDate
 	if !opts.Expiry.IsZero() {
 		expiresAt = jwt.NewNumericDate(opts.Expiry)
 	}
-	return &payload{
-		Notation: notationClaim{
-			Subject: desc,
-		},
+	return notaryClaim{
 		RegisteredClaims: jwt.RegisteredClaims{
-			IssuedAt:  jwt.NewNumericDate(time.Now()),
 			ExpiresAt: expiresAt,
+			IssuedAt:  jwt.NewNumericDate(time.Now()),
 		},
+		Notary: signature.JWSNotaryClaim{Subject: desc},
+	}
+}
+
+func checkCertChain(certChain []*x509.Certificate) error {
+	if len(certChain) == 0 {
+		return nil
+	}
+	if err := verifyCert(certChain[0], x509.ExtKeyUsageCodeSigning); err != nil {
+		return fmt.Errorf("signing certificate does not meet the minimum requirements: %w", err)
+	}
+	for _, c := range certChain[1:] {
+		for _, ext := range c.ExtKeyUsage {
+			if ext == x509.ExtKeyUsageTimeStamping {
+				if err := verifyCert(c, x509.ExtKeyUsageTimeStamping); err != nil {
+					return fmt.Errorf("timestamping certificate does not meet the minimum requirements: %w", err)
+				}
+			}
+		}
+	}
```

Comment:
It seems that we are having a timestamp CA in the cert chain. I'm a bit confused.

Reply:
The plugin spec doesn't say there can't be a timestamp CA in the certificate chain. It does says that the chain must be compliant with the [spec certificate requirements](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#certificate-requirements), which specifies the minimum requirements for a timestamp cert.

Reply:
Then I think it is an extra check. The cert requirement for timestamping should only apply to the timestamping signature and related certs.

Instead of having the discussion on the implementation here, we need to revisit the spec itself. @gokarnm Do you have any comments for this?

Reply:
I'll remove this check until the spec is clarified.

Reply:
Agree with Shiwei about the validation, and also that the spec is unclear. The certificate chain for primary signature (JWS `x5c`) needs to conform to signing certificate requirements. The timestamping cert requirements apply to cert chain in TSA countersignature (JWS `timestamp`). I think there is a gap here also for requirements for intermediate and root cert in the chain. 

@priteshbandi , the [cert requirements](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#certificate-requirements) section needs updates.

Reply:
Opened notaryproject/notaryproject#157 to track this.